### DISCUSSION TITLE
Secure Playback/Fairplay (Spotify and Apple Music)

### DISCUSSION THREAD

**Discussion by stevenboeren** [cefsharp/CefSharp#3601] [General]
Please ask your questions under the Discussions section here on GitHub 
- https://github.com/cefsharp/CefSharp/discussions

---

`CefSharp` is a wrapper around the [Chromium Embedded Framework](https://bitbucket.org/chromiumembedded/cef/overview) in a lot of cases a feature must be added to `CEF` first before it can be used in `CefSharp`. `CEF` has its own `Feature Request Forum` at https://magpcss.org/ceforum/viewforum.php?f=7

--- 

**Describe the solution you'd like**
A clear and concise description of the feature you are proposing

**Describe alternatives you've considered**

**Additional context**

**Checklist:**
- [ ] I have reviewed the [CEF API](https://magpcss.org/ceforum/apidocs3/index-all.html) and have confirmed the feature I'm requesting is possible.
- [ ] The Feature I'm requesting is an improvement to Async Javascript Binding (No new feature are being added to the Sync Javascript binding)
- [ ] An open `PR` exists on the [CEF Pull Requests](https://bitbucket.org/chromiumembedded/cef/pull-requests/) and I'd like to propose this feature is added to `CefSharp` when/if it's merged.

**Comment by amaitland** [comment]
They both require audio codecs that require royalties. See https://github.com/cefsharp/CefSharp/wiki/General-Usage#multimedia-audiovideo

Closing as this isn't something we can support out of the box. You have to build CEF from source.

**Comment by stevenboeren** [comment]
Hi Guys,

I wanted to make Apple Music program with CefSharp, so could use the program instead of webpage.
The website loads fine exept it will not load a song. After alot of reseaching yesterday I saw that you need to enable secured content (DRM). Because I tried Spotify web player and asks to enable secure playback. 

I was searching for hours and found something about WidevineCDM. But I am new to this and have no idea how it actually works. I will make a repo for this project if anyone wants to help.

Hope to hear from some of you soon!
Here are some screenshots of code and the errors.
![Code1](https://user-images.githubusercontent.com/70892306/120886064-4d041900-c5ec-11eb-93d0-9f4aaf4dc793.png)
![Code2](https://user-images.githubusercontent.com/70892306/120886066-5097a000-c5ec-11eb-9181-803a3eb40063.png)
![Screen1](https://user-images.githubusercontent.com/70892306/120885887-9011bc80-c5eb-11eb-9e07-851b844586c8.png)
![Screen2](https://user-images.githubusercontent.com/70892306/120885984-f72f7100-c5eb-11eb-936c-b1985c68ff15.png)

**Reply by amaitland** [reply]
Widevine DRM is already supported. 

- https://magpcss.org/ceforum/viewtopic.php?f=17&t=18419&p=49033#p49030
- http://cefsharp.github.io/api/90.6.x/search.html?SearchText=Widevine

The default builds don't include the AAC audio codec due to licensing restrictions. So once you enabled widevine you'll find the browser is still unable to play the audio. See https://magpcss.org/ceforum/viewtopic.php?f=6&t=17917&p=47013

**Reply by stevenboeren** [reply]
Well I found a delevoper webpage of Apple they use their own API.
https://developer.apple.com/documentation/musickitjs
https://developer.apple.com/forums/thread/113924

**Reply by amaitland** [reply]
If Apple music uses `AAC` audio codec then the audio won't play.

**Reply by stevenboeren** [reply]
Is there any way to make it play that? bc on Google Chrome it just works fine. I'm 17 and just wanna try somthing since I wanna be a Software Developer.

**Reply by amaitland** [reply]
Companies like Google and Microsoft pay licensing fees to include h264/AAC decoders in their products. CefSharp is open source, it only includes support for freely available codecs.

Your best bet is to try another browser, WebView2 from Microsoft might be a viable option.
